src/inet/mqtt_client.c

```diff
@@ -170,34 +218,153 @@ static int read_fixed_header(struct mqtt_client_t *self_p,
  */
 static int handle_control_connect(struct mqtt_client_t *self_p)
 {
-    int res = 0;
-    uint8_t buf[12];
+    struct mqtt_conn_options_t *options_p;
+    struct mqtt_conn_options_t default_options;
+    int res = 0, payload_length = 0;
+    uint8_t buf[CONNECT_VAR_HDR_LEN], flags = 0;
+
+    /*
+     * Note: Each payload string requires a 2 byte length header, so
+     * that must be accounted for in the payload length hence "+ 2" a
+     * number of length calculations below.
+     */
+
+    if (queue_read(&self_p->control.in,
+                   &options_p,
+                   sizeof(options_p)) != sizeof(options_p)) {
+        return (-1);
+    }
+
+    if (options_p == NULL) {
+        options_p = &default_options;
+        memset(options_p, 0, sizeof(*options_p));
+    }
+
+    /*
+     * We currently do not support ressuming session, so force clean
+     * session.
+     */
+    flags = CLEAN_SESSION;
+
+    /*
+      * Be sure that 'will' topic and payload are both either set or
+      * unset.
+      */
+    if ((options_p->will.topic.size == 0) !=
+        (options_p->will.payload.size == 0)) {
+        return (-EINVAL);
+    }
+
+    /*
+     * As per MQTT-3.1.3-3 a Client ID is required, so if the user has
+     * not specified one, we set one.
+     *
+     * We should preferably set a random client id, but we don't
+     * currently have an way to get a random number on all platforms.
+     */
+    if (options_p->client_id.size == 0) {
+        options_p->client_id.buf_p = FSTR("simba_mqtt");
```

Comment:
The `FSTR()` should only be used for strings of type `far_string_t`. Far strings are currently only used on AVR to allow strings to be stored in Flash instead of RAM.

I suggest removing the macro usage here, and keep the type of `buf_p` as `const char *`. The same change can be applied to other locations in this PR.

Reply:
Done.

Most of my time is still spend on computer with >128 GB of RAM, not <128 KB - so still haven't fully wrapped my head around some of the constraints here :-).